src/v/kafka/server/group_manager.cc

```diff
@@ -615,6 +623,64 @@ ss::future<> group_manager::reload_groups() {
     co_await ss::when_all_succeed(futures.begin(), futures.end());
 }
 
+ss::future<group_offsets_snapshot_result>
+group_manager::snapshot_groups(const model::ntp& ntp) {
+    auto attached_partition = _partitions.find(ntp);
+    if (attached_partition == _partitions.end()) {
+        co_return cluster::cloud_metadata::error_outcome::ntp_not_found;
+    }
+    // Avoid overlapping with concurrent reloads of the partition.
+    auto units = co_await ss::get_units(
+      attached_partition->second->sem, 1, attached_partition->second->as);
+    auto& catchup = attached_partition->second->catchup_lock;
+    co_await catchup->read_lock();
+    auto unlock = ss::defer([&catchup] { catchup->read_unlock(); });
+    if (!attached_partition->second->partition->is_leader()) {
+        co_return cluster::cloud_metadata::error_outcome::not_ready;
+    }
+    if (attached_partition->second->loading) {
+        co_return cluster::cloud_metadata::error_outcome::not_ready;
+    }
+    group_offsets_snapshot snap;
+    snap.offsets_topic_pid = ntp.tp.partition;
+
+    // Make a copy of the groups that we're about to snapshot, to avoid racing
+    // with removals during iteration.
+    fragmented_vector<std::pair<group_id, group_ptr>> groups;
+    std::copy_if(
+      _groups.begin(),
+      _groups.end(),
+      std::back_inserter(groups),
+      [&ntp](auto g_pair) {
+          const auto& [group_id, group] = g_pair;
+          return group->partition()->ntp().tp.partition == ntp.tp.partition;
+      });
+    vlog(klog.debug, "Snapshotting {} groups from {}", groups.size(), ntp);
+    for (const auto& [group_id, group] : groups) {
+        group_offsets go;
+        go.group_id = group_id();
+        absl::btree_map<
+          model::topic,
+          fragmented_vector<group_offsets::partition_offset>>
+          offsets;
+        for (const auto& [tp, o] : group->offsets()) {
+            offsets[tp.topic].emplace_back(
+              tp.partition, model::offset_cast(o->metadata.offset));
+        }
+        for (auto& [t, ps] : offsets) {
+            go.offsets.emplace_back(t, std::move(ps));
+        }
+        vlog(
+          klog.debug,
+          "Snapshotting offsets for {} topics from group {}",
+          go.offsets.size(),
+          go.group_id);
+        snap.groups.emplace_back(std::move(go));
+        co_await ss::maybe_yield();
+    }
+    co_return snap;
```

Comment:
It's a good suggestion. I don't think it's required because the snapshots themselves should be consistent, and if racing for the same key, the end result is still that the key exists by the end of it.

That said, maybe it would be a good failsafe to have just in case.

Reply:
I opened https://github.com/redpanda-data/redpanda/issues/15021 to track, and will discuss with more storage folks